### PR TITLE
Experiment: prototype of maybe terrible idea: navigation.dangerouslyUpdateRouter

### PR DESCRIPTION
**Pull request by brentvatne**
# Motivation

Some people would like to update the route config dynamically at runtime, eg: to add new routes to drawer or add/remove tabs. Before pursuing this further I'd like to make sure we understand and enumerate as many use cases as possible so we can know if this covers them. But I wanted to prototype this now to get an idea of the impact it might have on codebase.

# Test plan

It's a prototype, it doesn't have tests, you can open it in NavigationPlayground though.

### CODE REVIEW

File: src/navigators/createNavigator.js [hunk 1]
```diff
@@ -36,12 +37,23 @@ function createNavigator(NavigatorView, router, navigationConfig) {
       return this.props.navigation;
     };
 
+    _dangerouslyUpdateRouter = (key, newRouteConfigs) => {
+      if (router.setRoutes) {
```
Comment (ericvicenti) [inline src/navigators/createNavigator.js]: After the action creators overhaul, I don't think there would be a way to access the router from here.

Comment (brentvatne) [reply src/navigators/createNavigator.js]: this line is actually not needed because it's just using dispatch anyhow

**Comment by brentvatne** [comment]
@ericvicenti - agreed, I think that solution is nicer. I wonder if there are other use cases where updating the router would be useful. The main use cases I have in mind right now are adding routes to drawers / tabs, and removing routes from tabs.

If anyone has other cases where you need to use Redux to update your router state like we do here, please chime in here and let us know!

**Comment by jamonholmgren** [comment]
> Also, because routers are entirely stateless, I think it would be way easier to re-create the router if anything changes.

Any reason this shouldn't be the recommended way to approach this issue? I'm all for flexibility, but if it's adding complications to the implementation then it sounds like recreating the router is a reasonable approach.

**Comment by kevinsperrine** [comment]
In my snooping to try to find any-way-at-all to make dynamic routes work, I stumbled on this. 

@ericvicenti 
>Would it work to put "name" in params of FriendScreen, then use a custom contentComponent in the drawer, which asynchronously loads the names?
>In other words, what is the advantage of supporting .navigate('BrentScreen') if you can already do .navigate('FriendScreen', {name: 'Brent'})?

I don't see how listing the friends is possible without replacing large portions of the essentially core behavior in the custom drawer components. I've been trying to make dynamic routes work for weeks, and just keep running into different blockers and haven't been able to make it click. I highly appreciate all the work everyone puts into this library, but there are dozens of issues asking for dynamic routes, but nothing that _really_ shows all the ins-and-outs of making them work. 

In piecing together all the comments about custom components, setting static router props, just passing params, etc. I've started this snack. Can you explain what you mean about asynchronously loading the names and passing them into a custom component? Specially, how these new items can be added to the drawer as different items without running into descriptor errors or duplicate keys. 

See a minimal repo snack here: https://snack.expo.io/rk2ZCc2Jm

Thanks again for all the hard work.

**Pull request closed without merging**